### DISCUSSION TITLE
Epoptes Teacher Accounts Disconnect After Login

### DISCUSSION THREAD

**Discussion by robertmaynord** [epoptes/epoptes#100] [Q&A]
We have a school with Epoptes installed on all the student netbook.  It has worked fine for several years.  This year, the student machines show up on the server with no problems.  (version 1.0.1-2 Ubuntu 19.04).  The teachers have their usual accounts on the server and they can connect and see the student machines. It works fine for a short while. However, after a time between 30 seconds and 5 minutes the teachers are disconnected with the message: "Service Connection Error".  The server has a static IP, and the clients have the usual SERVER=192.168.0.20 in /etc/default/epoptes-client

In the teachers terminal, I then see the following error message....

Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1043, in _commandReceived
    deferred = self.dispatchCommand(box)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1101, in dispatchCommand
    return maybeDeferred(responder, box)
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1188, in doit
    return maybeDeferred(aCallable, **kw).addCallback(
--- <exception caught here> ---
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1171, in checkKnownErrors
    key = error.trap(*command.allErrors)
  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
    self.raiseException()
  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
    raise self.value.with_traceback(self.tb)
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/epoptes/daemon/uiconnection.py", line 43, in client_disconnected
    self.client.amp_client_disconnected(handle)
  File "/usr/lib/python3/dist-packages/epoptes/ui/gui.py", line 739, in amp_client_disconnected
    "host": client.get_name()})
  File "/usr/lib/python3/dist-packages/epoptes/ui/notifications.py", line 102, in enqueue
    self.show()
  File "/usr/lib/python3/dist-packages/epoptes/ui/notifications.py", line 135, in show
    self.notification.show()
gi.repository.GLib.Error: g-dbus-error-quark: Error calling StartServiceByName for org.freedesktop.Notifications: GDBus.Error:org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.Notifications exited with status 1 (25)

Error when executing command thumbshot 128 96 on client 192.168.0.117:35808: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/protocols/amp.py:1994:_massageError
/usr/lib/python3/dist-packages/twisted/python/failure.py:439:trap
/usr/lib/python3/dist-packages/twisted/python/failure.py:467:raiseException
]
Error when executing command thumbshot 128 96 on client 192.168.0.117:35734: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/protocols/amp.py:1994:_massageError
/usr/lib/python3/dist-packages/twisted/python/failure.py:439:trap
/usr/lib/python3/dist-packages/twisted/python/failure.py:467:raiseException
]
Error when executing command thumbshot 128 96 on client 192.168.0.127:37374: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/protocols/amp.py:1994:_massageError
/usr/lib/python3/dist-packages/twisted/python/failure.py:439:trap
/usr/lib/python3/dist-packages/twisted/python/failure.py:467:raiseException
]
Error when executing command thumbshot 128 96 on client 192.168.0.204:37902: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/protocols/amp.py:1994:_massageError
/usr/lib/python3/dist-packages/twisted/python/failure.py:439:trap
/usr/lib/python3/dist-packages/twisted/python/failure.py:467:raiseException
]
Error when executing command thumbshot 128 96 on client 192.168.0.214:44274: [Failure instance: Traceback: <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/protocols/amp.py:1994:_massageError
/usr/lib/python3/dist-packages/twisted/python/failure.py:439:trap
/usr/lib/python3/dist-packages/twisted/python/failure.py:467:raiseException
]
debian@slingshot-netbook:~$ 

Any suggestion as to where I might troubleshoot?

Robert Maynord

**Comment by alkisg** [comment]
Sounds like a problem with the notifications daemon.
Non LTS Ubuntu versions are really bad for schools and developers don't test them much, if at all.
It would be best to stick to 18.04 until 20.04 comes out.

That said, which desktop environment are you using?
And do the teachers login at the server locally, or remotely?

**Comment by robertmaynord** [comment]
Hi, Alkis!
I very much appreciate your help!
I am using the default desktop for Ubuntu, but even the old Epoptes netbooks seem to work fine.  The only problem is with the teachers.  They log into the server from their rooms.  It actually works fine for them as well - it just disconnects after a random period of time.  In my experimentation, I did find that if I log into the server and into a desktop - both at the same time, it seems to be OK until I disconnect on the server end - then the teacher disconnects.
It feels like it is very close to working if I could just discover the one tweak.....

Robert

**Comment by alkisg** [comment]
> They log into the server from their rooms. 

I do not know what that means. If the server is in a different room, which **software** are they using to connect? LTSP fat clients? LTSP thin clients? x2go? xfreerdp? VNC? ssh -X?

Notification issues are usually not a problem with Epoptes, but with the remote desktop software that people use.

**Comment by robertmaynord** [comment]
Indeed - the server is in the computer lab.  The teachers open a terminal on their classroom desktops, they type "ssh -X 192.168.0. 20 epoptes", then they type their password, and then epoptes appears on their screen.  All teachers are added to the group epoptes on the server (gpasswd -a username epoptes).  Perhaps I should try something besides ssh?

**Comment by alkisg** [comment]
This is a duplicate of issue #33,
see https://github.com/Epoptes/epoptes/issues/33#issuecomment-518976685

**Comment by robertmaynord** [comment]
Works!!!
